### PR TITLE
Feature/swaps

### PR DESCRIPTION
**Pull request by srindom**

### CODE REVIEW

File: packages/medusa/src/api/routes/admin/orders/archive-order.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ export default async (req, res) => {
   try {
     const orderService = req.scope.resolve("orderService")
     let order = await orderService.archive(id)
-    order = await orderService.decorate(order, [], ["region"])
+    order = await orderService.decorate(order, [], ["region", "swaps"])
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/archive-order.js:7]: Decorate with customer

File: packages/medusa/src/api/routes/admin/orders/cancel-order.js [hunk 2]
```diff
@@ -4,7 +4,7 @@ export default async (req, res) => {
   try {
     const orderService = req.scope.resolve("orderService")
     let order = await orderService.cancel(id)
-    order = await orderService.decorate(order, [], ["region"])
+    order = await orderService.decorate(order, [], ["region", "swaps"])
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/cancel-order.js:7]: decorate with customer

File: packages/medusa/src/api/routes/admin/orders/capture-payment.js [hunk 3]
```diff
@@ -4,7 +4,7 @@ export default async (req, res) => {
   try {
     const orderService = req.scope.resolve("orderService")
     let order = await orderService.capturePayment(id)
-    order = await orderService.decorate(order, [], ["region"])
+    order = await orderService.decorate(order, [], ["region", "swaps"])
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/capture-payment.js:7]: customer

File: packages/medusa/src/api/routes/admin/orders/capture-swap.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      order,
+      [],
+      ["region", "customer", "swaps"]
+    )
+
+    if (data.customer_id) {
+      const customerService = req.scope.resolve("customerService")
+      data.customer = await customerService.retrieve(order.customer_id)
+    }
+
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/capture-swap.js:23]: Move to decorater

File: packages/medusa/src/api/routes/admin/orders/complete-order.js [hunk 5]
```diff
@@ -4,7 +4,7 @@ export default async (req, res) => {
   try {
     const orderService = req.scope.resolve("orderService")
     let order = await orderService.completeOrder(id)
-    order = await orderService.decorate(order, [], ["region"])
+    order = await orderService.decorate(order, [], ["region", "swaps"])
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/complete-order.js:7]: customer

File: packages/medusa/src/api/routes/admin/orders/create-fulfillment.js [hunk 6]
```diff
@@ -26,8 +26,19 @@ export default async (req, res) => {
...
+    const data = await orderService.decorate(
+      order,
+      [],
+      ["region", "customer", "swaps"]
+    )
+
+    if (data.customer_id) {
+      const customerService = req.scope.resolve("customerService")
+      data.customer = await customerService.retrieve(order.customer_id)
+    }
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/create-fulfillment.js:39]: Remove

File: packages/medusa/src/api/routes/admin/orders/create-shipment.js [hunk 7]
```diff
@@ -22,7 +22,7 @@ export default async (req, res) => {
       value.fulfillment_id,
       value.tracking_numbers
     )
-    order = await orderService.decorate(order, [], ["region"])
+    order = await orderService.decorate(order, [], ["region", "swaps"])
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/create-shipment.js:25]: customer

File: packages/medusa/src/api/routes/admin/orders/create-swap-shipment.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    const data = await orderService.decorate(
+      order,
+      [],
+      ["region", "customer", "swaps"]
+    )
+
+    if (data.customer_id) {
+      const customerService = req.scope.resolve("customerService")
+      data.customer = await customerService.retrieve(order.customer_id)
+    }
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/create-swap-shipment.js:40]: Remove

File: packages/medusa/src/api/routes/admin/orders/get-order.js [hunk 9]
```diff
@@ -6,7 +6,11 @@ export default async (req, res) => {
...
     let order = await orderService.retrieve(id)
-    order = await orderService.decorate(order, [], ["region", "customer"])
+    order = await orderService.decorate(
+      order,
+      [],
+      ["region", "customer", "swaps"]
+    )
 
     if (order.customer_id) {
       order.customer = await customerService.retrieve(order.customer_id)
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/get-order.js:16]: Remove

File: packages/medusa/src/api/routes/admin/orders/index.js [hunk 10]
```diff
@@ -95,6 +95,43 @@ export default app => {
     middlewares.wrap(require("./archive-order").default)
   )
 
+  /**
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/index.js:98]: Add cancel swap

File: packages/medusa/src/api/routes/admin/orders/receive-return.js [hunk 11]
```diff
@@ -25,14 +25,14 @@ export default async (req, res) => {
...
-    let order = await orderService.return(
+    let order = await orderService.receiveReturn(
       id,
       return_id,
       value.items,
       refundAmount,
       true
     )
-    order = await orderService.decorate(order, [], ["region"])
+    order = await orderService.decorate(order, [], ["region", "swaps"])
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/receive-return.js:35]: Customer

File: packages/medusa/src/api/routes/admin/orders/receive-swap.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    const data = await orderService.decorate(
+      order,
+      [],
+      ["region", "customer", "swaps"]
+    )
+
+    if (data.customer_id) {
+      const customerService = req.scope.resolve("customerService")
+      data.customer = await customerService.retrieve(order.customer_id)
+    }
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/receive-swap.js:43]: Remove

File: packages/medusa/src/api/routes/admin/orders/refund-payment.js [hunk 13]
```diff
@@ -23,7 +23,7 @@ export default async (req, res) => {
       value.reason,
       value.note
     )
-    order = await orderService.decorate(order, [], ["region"])
+    order = await orderService.decorate(order, [], ["region", "swaps"])
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/refund-payment.js:26]: Customer

File: packages/medusa/src/api/routes/admin/orders/request-return.js [hunk 14]
```diff
@@ -66,7 +66,7 @@ export default async (req, res) => {
       )
     }
 
-    order = await orderService.decorate(order, [], ["region"])
+    order = await orderService.decorate(order, [], ["region", "swaps"])
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/request-return.js:69]: Customer

File: packages/medusa/src/api/routes/admin/orders/set-metadata.js [hunk 15]
```diff
@@ -16,7 +16,7 @@ export default async (req, res) => {
   try {
     const orderService = req.scope.resolve("orderService")
     let order = await orderService.setMetadata(id, value.key, value.value)
-    order = await orderService.decorate(order, [], ["region"])
+    order = await orderService.decorate(order, [], ["region", "swaps"])
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/set-metadata.js:19]: Customer

File: packages/medusa/src/api/routes/admin/orders/update-order.js [hunk 16]
```diff
@@ -32,7 +32,7 @@ export default async (req, res) => {
   try {
     const orderService = req.scope.resolve("orderService")
     let order = await orderService.update(id, value)
-    order = await orderService.decorate(order, [], ["region"])
+    order = await orderService.decorate(order, [], ["region", "swaps"])
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/orders/update-order.js:35]: Customer

File: packages/medusa/src/api/routes/store/swaps/index.js [hunk 17]
```diff
@@ -0,0 +1,13 @@
+import { Router } from "express"
+import middlewares from "../../../middlewares"
+
+const route = Router()
+
+export default app => {
+  app.use("/swaps", route)
+
+  // route.get("/:id", middlewares.wrap(require("./get-swap").default))
```
Comment (srindom) [inline packages/medusa/src/api/routes/store/swaps/index.js:9]: Add get swap route

File: packages/medusa/src/models/swap.js [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+import ReturnSchema from "./schemas/return"
+import FulfillmentSchema from "./schemas/fulfillment"
+import PaymentMethodSchema from "./schemas/payment-method"
+import ShippingMethodSchema from "./schemas/shipping-method"
+import AddressSchema from "./schemas/address"
+
+class SwapModel extends BaseModel {
+  static modelName = "Swap"
+  static schema = {
+    status: { type: String, default: "pending" },
```
Comment (srindom) [inline packages/medusa/src/models/swap.js:14]: Not being used

File: packages/medusa/src/services/cart.js [hunk 19]
```diff
@@ -208,10 +208,12 @@ class CartService extends BaseService {
...
 
     const region = await this.regionService_.retrieve(region_id)
-    if (region.countries.length === 1) {
-      // Preselect the country if the region only has 1
-      data.shipping_address = {
-        country_code: region.countries[0],
+    if (!data.shipping_address) {
+      if (region.countries.length === 1) {
+        // Preselect the country if the region only has 1
+        data.shipping_address = {
```
Comment (srindom) [inline packages/medusa/src/services/cart.js:214]: Add a test for this

File: packages/medusa/src/services/order.js [hunk 20]
```diff
@@ -1422,6 +1238,87 @@ class OrderService extends BaseService {
...
   }
 
+  /**
+   * Registers a swap to the order. The swap must belong to the order for it to
+   * be registered.
+   * @param {string} id - the id of the order to register the swap to.
+   * @param {string} swapId - the id of the swap to add to the order.
+   * @returns {Promise<Order>} the resulting order
+   */
+  async registerSwapCreated(id, swapId) {
```
Comment (srindom) [inline packages/medusa/src/services/order.js:1248]: Add test for failure points

File: packages/medusa/src/services/order.js [hunk 21]
```diff
@@ -1422,6 +1238,87 @@ class OrderService extends BaseService {
...
+  }
+
+  /**
+   * Registers the swap return items as received so that they cannot be used
+   * as a part of other swaps/returns.
+   * @param {string} id - the id of the order with the swap.
+   * @param {string} swapId - the id of the swap that has been received.
+   * @returns {Promise<Order>} the resulting order
+   */
+  async registerSwapReceived(id, swapId) {
```
Comment (srindom) [inline packages/medusa/src/services/order.js:1277]: Add test for failure points

File: packages/medusa/src/services/order.js [hunk 22]
```diff
@@ -1422,6 +1238,87 @@ class OrderService extends BaseService {
...
+        const returned = returnedQuantity === i.quantity
+        return {
+          ...i,
+          returned_quantity: returnedQuantity,
+          returned,
+        }
+      }
+
+      return i
+    })
```
Comment (srindom) [inline packages/medusa/src/services/order.js:1309]: Add test for logic

File: packages/medusa-plugin-brightpearl/src/services/brightpearl.js [hunk 23]
```diff
@@ -471,6 +472,190 @@ class BrightpearlService extends BaseService {
...
+    if (!soId) {
+      return
+    }
+
+    const paymentType = "RECEIPT"
+    const paymentMethod = fromSwap.payment_method
+    const payment = {
+      transactionRef: `${paymentMethod._id}.${paymentType}`, // Brightpearl cannot accept an auth and capture with same ref
+      transactionCode: fromSwap._id,
+      paymentMethodCode: "1220",
```
Comment (olivermrbl) [inline packages/medusa-plugin-brightpearl/src/services/brightpearl.js:489]: Where does this number come from?

File: packages/medusa/src/api/routes/admin/customers/get-customer.js [hunk 24]
```diff
@@ -12,9 +13,14 @@ export default async (req, res) => {
         "shipping_addresses",
         "phone",
       ],
-      ["orders"]
+      []
     )
 
+    customer.orders = await Promise.all(customer.orders.map(async oId => {
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/customers/get-customer.js:19]: Move this to `customerService.decorate` right? We want to be consistent.

File: packages/medusa/src/api/routes/admin/customers/update-customer.js [hunk 25]
```diff
@@ -16,11 +16,18 @@ export default async (req, res) => {
   }
 
   try {
+    const orderService = req.scope.resolve("orderService")
     const customerService = req.scope.resolve("customerService")
     await customerService.update(id, value)
 
     const customer = await customerService.retrieve(id)
     const data = await customerService.decorate(customer)
+    data.orders = await Promise.all(
```
Comment (olivermrbl) [inline packages/medusa/src/api/routes/admin/customers/update-customer.js:25]: See comment above.

File: packages/medusa/src/services/return.js [hunk 26]
```diff
@@ -0,0 +1,293 @@
...
+        item_id: i._id,
+        content: i.content,
+        quantity: i.quantity,
+        is_requested: true,
+        metadata: i.metadata,
+      })),
+      shipping_data: fulfillmentData,
+    }
+
+    //return this.orderModel_
```
Comment (olivermrbl) [inline packages/medusa/src/services/return.js:188]: Remove

**Review by olivermrbl — CHANGES_REQUESTED** [review]
Amazing job! 🏅 

We should consider adding a `DecoratorService` at some point?

**Pull request merged**